c10/util/Metaprogramming.h

```diff
@@ -17,12 +18,24 @@ namespace c10 { namespace guts {
 template<class Func> struct function_traits {
   static_assert(!std::is_same<Func, Func>::value, "In function_traits<Func>, Func must be a plain function type.");
 };
-template<class Result, class... Args>
-struct function_traits<Result (Args...)> {
-  using func_type = Result (Args...);
+template<class Result>
+struct function_traits<Result ()> {
+  using func_type = Result ();
   using return_type = Result;
-  using parameter_types = typelist::typelist<Args...>;
-  static constexpr auto number_of_parameters = sizeof...(Args);
+  using parameter_types = typelist::typelist<>;
+  static constexpr auto number_of_parameters = 0;
+  using first_arg_is_dispatchKeySet = std::is_same<DispatchKeySet, Result>;
+};
+
+template<class Result, class FirstArg, class... Args>
```

Comment:
This specialization could also be removed again if we eliminate the dependency to DispatchKey from here.

Reply:
I don't currently see a way to remove it easily without exposing the `DispatchKeySet` type to jit (see my comment https://github.com/pytorch/pytorch/pull/49354#discussion_r555962009)

Reply:
I don't follow. If you just have the function_traits above without that specialization and use `head_with_default_t` on `parameter_types`, then there is no reason to specialize function_traits anymore or am I missing something?